src/main/java/ch/njol/skript/effects/EffTeleport.java

```diff
@@ -19,125 +19,141 @@
 package ch.njol.skript.effects;
 
 import ch.njol.skript.Skript;
-import ch.njol.skript.sections.EffSecSpawn;
-import ch.njol.skript.sections.EffSecSpawn.SpawnEvent;
-import ch.njol.skript.bukkitutil.EntityUtils;
+import ch.njol.skript.bukkitutil.TeleportFlags;
+import ch.njol.skript.bukkitutil.TeleportFlags.SkriptTeleportFlags;
 import ch.njol.skript.doc.Description;
 import ch.njol.skript.doc.Examples;
 import ch.njol.skript.doc.Name;
+import ch.njol.skript.doc.RequiredPlugins;
 import ch.njol.skript.doc.Since;
 import ch.njol.skript.lang.Effect;
 import ch.njol.skript.lang.Expression;
 import ch.njol.skript.lang.SkriptParser.ParseResult;
 import ch.njol.skript.lang.Trigger;
 import ch.njol.skript.lang.TriggerItem;
-import ch.njol.skript.lang.TriggerSection;
+import ch.njol.skript.sections.EffSecSpawn.SpawnEvent;
 import ch.njol.skript.timings.SkriptTimings;
 import ch.njol.skript.util.Direction;
 import ch.njol.skript.variables.Variables;
 import ch.njol.util.Kleenean;
+
 import io.papermc.lib.PaperLib;
 import io.papermc.lib.environments.PaperEnvironment;
+
 import org.bukkit.Location;
 import org.bukkit.entity.Entity;
 import org.bukkit.event.Event;
 import org.bukkit.event.player.PlayerMoveEvent;
 import org.bukkit.event.player.PlayerRespawnEvent;
 import org.eclipse.jdt.annotation.Nullable;
+import org.jetbrains.annotations.NotNull;
 
 @Name("Teleport")
 @Description({
 	"Teleport an entity to a specific location. ",
 	"This effect is delayed by default on Paper, meaning certain syntax such as the return effect for functions cannot be used after this effect.",
 	"The keyword 'force' indicates this effect will not be delayed, ",
-	"which may cause lag spikes or server crashes when using this effect to teleport entities to unloaded chunks."
+	"which may cause lag spikes or server crashes when using this effect to teleport entities to unloaded chunks.",
+	"Teleport flags are settings to retain during a teleport. Such as direction, passengers, x coordinate, etc."
 })
 @Examples({
-	"teleport the player to {homes.%player%}",
-	"teleport the attacker to the victim"
+	"teleport the player to {home::%uuid of player%}",
+	"teleport the attacker to the victim",
+	"",
+	"on dismount:",
+		"\tcancel event",
+		"\tteleport the player to {server::spawn} retaining vehicle and passengers"
 })
-@Since("1.0")
+@RequiredPlugins("Paper 1.19+ (teleport flags)")
+@Since("1.0, INSERT VERSION")
 public class EffTeleport extends Effect {
 
+	private static final boolean TELEPORT_FLAGS = Skript.classExists("io.papermc.paper.entity.TeleportFlag");
 	private static final boolean CAN_RUN_ASYNC = PaperLib.getEnvironment() instanceof PaperEnvironment;
 
 	static {
-		Skript.registerEffect(EffTeleport.class, "[(1¦force)] teleport %entities% (to|%direction%) %location%");
+		String extra = "";
+		if (TELEPORT_FLAGS)
+			extra = " [[retaining] %teleportflags%]";
```

Comment:
The language nodes contain the word `retain` this `retaining` is for allowing multiples `retaining direction, passengers and vehicles`

I like the `while` addition though.

Reply:
Sadly it's also allowed to do `retaining retain direction, retain passengers` so that's bad therefore we need to either enforce `retain` keyword in language nodes (which is preferred because we don't want to have conflicting types like `inventory, x, y, z, direction, vehicle, passenger`) or remove `retaining` but as said the former is very recommended.

We can simply add optional `ing` to types like so `retain[ing] open inventory` this will solve everything.

Reply:
There are many of these possibilities in Skript. I don't want to go with this change. Including the `retain` in the default.lang makes it easier to distinguish between others for the parser, and retaining for multiples.